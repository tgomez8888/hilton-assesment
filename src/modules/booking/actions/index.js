import ActionType from "../constants/actionType";

export function setRoomsValue(rooms) {
  return {
    type: ActionType.SET_ROOMS_VALUE,
    payload: { rooms }
  };
}

export function roomValueChange(id, name, value) {
  return {
    type: ActionType.ROOM_VALUE_CHANGE,
    payload: { id, name, value }
  };
}

export function roomSelectionChange(id, name, value) {  
  return {
    type: ActionType.ROOM_SELECTION_CHANGE,
    payload: { id, name, value }
  };
}
