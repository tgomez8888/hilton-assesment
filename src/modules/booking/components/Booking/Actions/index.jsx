import React from "react";

export function Actions() {
  return (
    <div className="actions-panel">
      <button type="submit">Submit</button>
    </div>
  );
}

export default Actions;
