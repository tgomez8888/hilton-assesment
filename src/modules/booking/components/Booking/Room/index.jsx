import React from "react";
import classnames from "classnames";
import RoomHeader from "./RoomHeader";
import RoomBody from "./RoomBody";
import enhance from "./enhancers";

export function Room({ room, onRoomSelectedChange, onRoomValueChange }) {
  const roomClasses = classnames({
    "card-container": true,
    "disabled": !room.selected
  });

  return (
    <div className={roomClasses}>
      <RoomHeader room={room} onRoomValuesChange={onRoomSelectedChange} />
      <RoomBody room={room} onRoomValuesChange={onRoomValueChange} />
    </div>
  );
}

export default enhance(Room);
