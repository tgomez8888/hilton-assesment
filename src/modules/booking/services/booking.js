import {
  __,
  always,
  assoc,
  curry,
  evolve,
  gte,
  ifElse,
  identity,
  indexBy,
  lt,
  map,
  pipe,
  prop,
  values
} from "ramda";

export const defaultRooms = [
  { id: 1, priority: 1, selected: true, adults: "1", children: "0" },
  { id: 2, priority: 2, selected: false, adults: "1", children: "0" },
  { id: 3, priority: 3, selected: false, adults: "1", children: "0" },
  { id: 4, priority: 4, selected: false, adults: "1", children: "0" }
];

export const updateRoom =  curry((newRoom, rooms) => pipe(
  indexBy(prop("id")),
  assoc(newRoom.id, newRoom),
  values
)(rooms)); 

const isHigherPriority = curry((newRoom, room) => pipe(
  prop("priority"),
  lt(__, newRoom.priority)
)(room));

const isLowerOrSamePriority = curry((newRoom, room) => pipe(
  prop("priority"),
  gte(__, newRoom.priority)
)(room));

const setDefaultValues = evolve({
  selected: always(false),
  adults: always("1"),
  children: always("0")
});

/**
 * This method receives a room with changes and calculates if the other rooms need to changes.
 * If room is selected all rooms with higher priority (lower priority value) will 
 * be selected.
 * If room is not selected all rooms with lower or same priority (higher priority value) will be 
 * unselected and set default values.
 * @param {Object} newRoom This is the modified room object 
 * @param {[Object]} rooms The current list of rooms in state
 * @returns {[Object]} New list of rooms with modifications 
 */
export const processRoomSelectionChanges = (newRoom, rooms) => {
   
  const turnOnHigherPriority = map(
    ifElse(isHigherPriority(newRoom), assoc("selected", true), identity)
  ); 
  const turnOffLowerPriority = map(
    ifElse(
      isLowerOrSamePriority(newRoom),
      setDefaultValues,
      identity
    )
  );

  return  pipe(
    updateRoom(newRoom),
    ifElse(() => newRoom.selected, turnOnHigherPriority, turnOffLowerPriority)
  )(rooms);
}