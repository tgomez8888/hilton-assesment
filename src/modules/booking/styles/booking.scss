@import "./modules/variables";

body {
	font-size: 14px;
}

input, select, button {
	font-size: 14px;
	margin: 0;
}

.actions-panel {
	padding: 10px;
}

.card-container {
	background-color: $color-background;
	border: 4px solid $color-border;
	display: inline-block;
	margin: 10px;
	border-radius: 5%;
	.card-header {
		background-color: $color-background;
		font-weight: bold;
		padding: 5px;
	}
	.card-body {
		background-color: $color-background-content;
		padding: 5px;
	}
	.column {
		display: inline-block;
		padding-right: 5px;
	}	
}

.card-container.disabled {
	background-color: $color-background-disabled;
	border-color: $color-border-disabled;

	.card-header {
		background-color: $color-background-disabled;
		font-weight: normal;
	}

	.card-body {
		background-color: $color-background-disabled;
	}	
}

.guess-selector {
	label{
		font-size: 12px;		
	}	
	width: 50px;	
}