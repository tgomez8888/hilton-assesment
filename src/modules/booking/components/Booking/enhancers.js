import { connect } from "react-redux";
import { compose, lifecycle } from "recompose";
import { either, equals, isEmpty, isNil} from "ramda";
import { getRoomsId, getRooms } from "../../selectors";
import {setRoomsValue} from "../../actions";

const mapDispatchToProps = {
  setRoomsValue
};
function mapStateToProps(state, {setRoomsValue}) {
  const rooms = getRooms(state);
  const onSubmit = e => {
    e.preventDefault();
    // Here you could call GraphQL mutation to save rooms to server
    console.log(rooms);
  };
  return {
		ids: getRoomsId(state),
		onLoadRooms: setRoomsValue,
		onSubmit
  };
}

export default compose(
  // GraphQL Query to get rooms saved Data or load them in props
  connect(
    null,
    mapDispatchToProps
  ),
  lifecycle({
		componentDidMount(){
			const {rooms, setRoomsValue} = this.props;
			
			// If component is loaded with a value of rooms in props load it in state 
			if(!either(isNil, isEmpty)(rooms)){
				setRoomsValue(rooms);
			}
		},
		componentWillReaceiveProps(newProps){
			// If rooms are changed in components props to something different, update state.
			if(newProps.rooms && !equals(newProps.rooms,this.props.rooms)){
				this.props.setRoomsValue(newProps.rooms);
			}
		}
	}),
  connect(mapStateToProps)
);
