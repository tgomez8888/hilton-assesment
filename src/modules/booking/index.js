import Booking from "./components/Booking";

export default {Booking};