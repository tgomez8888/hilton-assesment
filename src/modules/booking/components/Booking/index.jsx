import React, { Component } from "react";
import { either, equals, ifElse, identity, isEmpty, isNil } from "ramda";
import {
  defaultRooms,
  processRoomSelectionChanges,
  updateRoom
} from "../../services/booking";
import Room from "./Room";
import Actions from "./Actions";
import "../../styles/booking.scss";

export class Booking extends Component {
  constructor(props) {
    super(props);

    this.state = {
      rooms: ifElse(either(isNil, isEmpty), () => defaultRooms, identity)(
        props.rooms
      )
    };
  }

  componentWillReceiveProps(newProps) {
    // I would expect an HOC that wraps this component and calls a GraphQL Query to get the rooms
    // and pass the rooms value as a prop either initially or as a load function. This
    // will be for load function, constructor is for initial prop load

    if (newProps.rooms && !equals(newProps.rooms, this.props.rooms)) {
      this.setState({ rooms: newProps.rooms });
    }
  }

  handleRoomValueChanges = newRoom => {
    this.setState({ rooms: updateRoom(newRoom, this.state.rooms) });
  };

  handleRoomSelectionChanges = newRoom => {
    this.setState({
      rooms: processRoomSelectionChanges(newRoom, this.state.rooms)
    });
  };

  handleSubmit = e => {
    e.preventDefault();
    /* Here we can call the GraphQL mutation that sends the rooms list stored in this.state.rooms */

    console.log(this.state.rooms);
  };

  render() {
    const { rooms } = this.state;
    return (
      <main>
        <form onSubmit={this.handleSubmit}>
          {rooms &&
            rooms.map(room => (
              <Room
                key={room.id}
                room={room}
                onRoomValuesChange={this.handleRoomValueChanges}
                onRoomSelectionChange={this.handleRoomSelectionChanges}
              />
            ))}
          <Actions />
        </form>
      </main>
    );
  }
}

export default Booking;
