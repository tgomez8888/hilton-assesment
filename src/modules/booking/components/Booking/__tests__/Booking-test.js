import React from "react";
import { mount } from "enzyme";
import  { defaultRooms } from "../../../services/booking";
import Booking from "..";

describe("Booking", () => {
  function render({ rooms }) {
    return mount(<Booking rooms={rooms} />);
  }

  test("renders corectly", () => {
    const wrapper = render({});

    expect(wrapper.html()).toMatchSnapshot();
  });

  test("when rooms are passed as props they override the state value", () => {
    const initialRooms = [
      { id: 1, priority: 1, selected: true, adults: "2", children: "1" },
      { id: 2, priority: 2, selected: true, adults: "1", children: "2" },
      { id: 3, priority: 3, selected: false, adults: "1", children: "0" },
      { id: 4, priority: 4, selected: false, adults: "1", children: "0" }
    ];
    const wrapper = render({ rooms: initialRooms });

    expect(wrapper.state().rooms).toEqual(initialRooms);
  });

  test("when no rooms passed as props the state uses default", () => {
    const wrapper = render({});

    expect(wrapper.state().rooms).toEqual(defaultRooms);
  });

  //   function testChangedValue(fieldName, value) {
  //     const wrapper = render();

  //     wrapper
  //       .find("[name='" + fieldName + "']")
  //       .first()
  //       .simulate("change", { target: { name: fieldName, value } });

  //   }

  //   function testChangedChecked(fieldName, value) {
  //     const testFunction = jest.fn();
  //     const room = {
  //       id: 1,
  //       priority: 2,
  //       selected: false,
  //       adults: 1,
  //       children: 0
  //     };
  //     const onChange = obj => {
  //       expect(differenceObjects(obj, room)).toEqual([[fieldName, value]]);
  //       testFunction(obj);
  //     };
  //     const wrapper = renderEnhanced({
  //       room,
  //       onChange
  //     });

  //     wrapper
  //       .find("[name='" + fieldName + "']")
  //       .first()
  //       .simulate("change", { target: { name: fieldName, checked: value } });

  //     expect(testFunction).toBeCalled();
  //   }

  //   test("when selected is changed the onchange is called with a new value in selected", () => {
  //     testChangedChecked("selected", true);
  //   });

  //   test("when selected is changed the onchange is called with a new value in adults", () => {
  //     testChangedValue("adults", "2");
  //   });

  //   test("when selected is changed the onchange is called with a new value in children", () => {
  //     testChangedValue("children", "1");
  //   });
});
