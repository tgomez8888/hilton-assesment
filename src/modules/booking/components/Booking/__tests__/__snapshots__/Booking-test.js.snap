// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking renders corectly 1`] = `"<main><form><div class=\\"card-container\\"><div class=\\"card-header\\">Room 1</div><div class=\\"card-body\\"><div class=\\"column guess-selector\\"><label>Adults (18+)</label><select name=\\"adults\\"><option value=\\"1\\">1</option><option value=\\"2\\">2</option></select></div><div class=\\"column guess-selector\\"><label>Children (0-17)</label><select name=\\"children\\"><option value=\\"0\\">0</option><option value=\\"1\\">1</option><option value=\\"2\\">2</option></select></div></div></div><div class=\\"card-container disabled\\"><div class=\\"card-header\\"><input type=\\"checkbox\\" name=\\"selected\\">Room 2</div><div class=\\"card-body\\"><div class=\\"column guess-selector\\"><label>Adults (18+)</label><select name=\\"adults\\" disabled=\\"\\"><option value=\\"1\\">1</option><option value=\\"2\\">2</option></select></div><div class=\\"column guess-selector\\"><label>Children (0-17)</label><select name=\\"children\\" disabled=\\"\\"><option value=\\"0\\">0</option><option value=\\"1\\">1</option><option value=\\"2\\">2</option></select></div></div></div><div class=\\"card-container disabled\\"><div class=\\"card-header\\"><input type=\\"checkbox\\" name=\\"selected\\">Room 3</div><div class=\\"card-body\\"><div class=\\"column guess-selector\\"><label>Adults (18+)</label><select name=\\"adults\\" disabled=\\"\\"><option value=\\"1\\">1</option><option value=\\"2\\">2</option></select></div><div class=\\"column guess-selector\\"><label>Children (0-17)</label><select name=\\"children\\" disabled=\\"\\"><option value=\\"0\\">0</option><option value=\\"1\\">1</option><option value=\\"2\\">2</option></select></div></div></div><div class=\\"card-container disabled\\"><div class=\\"card-header\\"><input type=\\"checkbox\\" name=\\"selected\\">Room 4</div><div class=\\"card-body\\"><div class=\\"column guess-selector\\"><label>Adults (18+)</label><select name=\\"adults\\" disabled=\\"\\"><option value=\\"1\\">1</option><option value=\\"2\\">2</option></select></div><div class=\\"column guess-selector\\"><label>Children (0-17)</label><select name=\\"children\\" disabled=\\"\\"><option value=\\"0\\">0</option><option value=\\"1\\">1</option><option value=\\"2\\">2</option></select></div></div></div><div class=\\"actions-panel\\"><button type=\\"submit\\">Submit</button></div></form></main>"`;
